torch_geometric/nn/conv/peg_conv.py

```diff
@@ -0,0 +1,196 @@
+    `"Equivariant and Stable Positional Encoding
+    for More Powerful Graph Neural Networks"
+    <https://arxiv.org/abs/2203.00199>`_ paper
+
+
+    Args:
+        in_feats_dim (int): Size of input node features.
+        pos_dim (int): Size of positional encoding.
+        out_feats_dim (int): Size of output node features.
+        edge_mlp_dim (int):
+            We use MLP to make one to one mapping between the
+            relative information and edge weight. edge_mlp_dim
+            represents the hidden units dimension in the MLP. (default: 32)
+        improved (bool, optional): If set to :obj:`True`, the layer computes
+            :math:`\mathbf{\hat{A}}` as :math:`\mathbf{A} + 2\mathbf{I}`.
+            (default: :obj:`False`)
+        cached (bool, optional): If set to :obj:`True`, the layer will cache
+            the computation of :math:`\mathbf{\hat{D}}^{-1/2} \mathbf{\hat{A}}
+            \mathbf{\hat{D}}^{-1/2}` on first execution, and will use the
+            cached version for further executions.
+            This parameter should only be set to :obj:`True` in transductive
+            learning scenarios. (default: :obj:`False`)
+        add_self_loops (bool, optional): If set to :obj:`False`, will not add
+            self-loops to the input graph. (default: :obj:`True`)
+        normalize (bool, optional): Whether to add self-loops and compute
+            symmetric normalization coefficients on the fly.
+            (default: :obj:`True`)
+        bias (bool, optional): If set to :obj:`False`, the layer will not learn
+            an additive bias. (default: :obj:`True`)
+        use_formerinfo (bool):
+            Whether to use previous layer's output to update node features.
+        **kwargs (optional): Additional arguments of
+            :class:`torch_geometric.nn.conv.MessagePassing`.
+    """
+    def __init__(self, in_feats_dim: int, pos_dim: int, out_feats_dim: int,
+                 edge_mlp_dim: int = 32, improved: bool = False,
+                 cached: bool = False, add_self_loops: bool = True,
+                 normalize: bool = True, bias: bool = True,
+                 use_formerinfo: bool = False, **kwargs):
+
+        kwargs.setdefault('aggr', 'add')
+        super(PEGConv, self).__init__(**kwargs)
+
+        self.in_feats_dim = in_feats_dim
+        self.out_feats_dim = out_feats_dim
+        self.pos_dim = pos_dim
+        self.use_formerinfo = use_formerinfo
+        self.improved = improved
+        self.cached = cached
+        self.add_self_loops = add_self_loops
+        self.normalize = normalize
+        self.edge_mlp_dim = edge_mlp_dim
+
+        self._cached_edge_index = None
+        self._cached_adj_t = None
+
+        self.weight_withformer = Parameter(
+            torch.Tensor(in_feats_dim + in_feats_dim, out_feats_dim))
+        self.weight_noformer = Parameter(
+            torch.Tensor(in_feats_dim, out_feats_dim))
+        self.edge_mlp = nn.Sequential(nn.Linear(1, edge_mlp_dim),
```

Comment:
[newbie question] somewhat confused by this - If you go from 1 dim back to 1 dim are you able learning much?

Reply:
Maybe I can use 'information bottleneck' to explain it. Though I go from 1 dim back to 1 dim, the network rids noisy input data of extraneous details as if by squeezing the information through a bottleneck, retaining only the features most relevant to the edge weight.